ykrt/src/compile/jitc_yk/jit_ir/mod.rs

```diff
@@ -1594,9 +1608,59 @@ impl PtrAddInst {
         ptr.unpack()
     }
 
-    pub(crate) fn offset(&self) -> Operand {
-        let unaligned = std::ptr::addr_of!(self.off);
-        unsafe { std::ptr::read_unaligned(unaligned) }.unpack()
+    pub(crate) fn off(&self) -> i32 {
+        self.off
+    }
+}
+
+/// An instruction that adds a (potentially) dynamic offset to a pointer.
+///
+/// The dynamic value is computed from an element size and a number of elements.
+///
+/// # Semantics
+///
+/// Returns a pointer value that is the result of:
+///  - multiplying the constant element size by the dynamic number of elements, and
+///  - adding the result to the specified pointer.
+///
+/// Following LLVM semantics, the operation is permitted to silently wrap if the result doesn't fit
+/// in the LLVM pointer indexing type.
+#[derive(Clone, Debug, PartialEq)]
+#[repr(packed)]
+pub struct DynPtrAddInst {
+    /// The pointer to offset
+    ptr: PackedOperand,
+    /// The element size.
+    ///
+    /// Depending upon the platform, LLVM `getelementptr` may allow larger element sizes than what
+    /// this field can express. Due to space constraints we only accept a reduced range of values
+    /// and traces requiring values outside of this range cannot be JIT compiled.
+    elem_size: u16,
+    /// The (dynamic) number of elements.
+    num_elems: PackedOperand,
```

Comment:
If I understand correctly, the way to implement `DynPtrAdd` is (roughly) `m.type_(self.ptr).byte_size() * self.num_elems`?

Reply:
I was asking a question not making a statement, because I don't know the answer :) I guess your question implies the answer to my question is "no"?

Reply:
No, you can't assume that array elements are pointer-sized.

This is why in the LLVM API [GetElementPtrInst::collectOffset()](https://llvm.org/doxygen/classllvm_1_1GetElementPtrInst.html#a600fb2213a893d3db4d03ea9fe6b9e63), returns a vector of dynamic indices that are pairs of `Value, APInt`. The former is the dynamic number of elements, and the `APInt`s the constant byte-size of the elements of the array being indexed.

Then in the [language reference](https://llvm.org/docs/LangRef.html#id237) they explain:

> the index itself is sign-extended or truncated, and then multiplied by the type allocation size (that is, the size rounded up to the ABI alignment) of the currently indexed type. The offsets are then added to the low bits of the base address up to the index type width, with silently-wrapping two’s complement arithmetic.

A lot of detail, but I hope it roughly makes sense?